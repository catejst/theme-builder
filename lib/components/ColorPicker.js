import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { CustomPicker, SketchPicker } from 'react-color';
import { EditableInput } from 'react-color/lib/components/common';
import { defaultColor } from '../utils/consts';

const ColorPicker = (props) => {
  const { title, initialColor, onChange } = props;
  const defaultColorObj = { hex: defaultColor };
  const [color, setColor] = useState(defaultColorObj);
  const [displayColorPicker, setColorPickerDisplay] = useState(false);
  useEffect(() => {
    if (initialColor) {
      setColor({ hex: initialColor });
    }
  }, []);
  const styles = {
    input: {
      border: 'none',
      outline: 'none',
      fontSize: '14px'
    },
    label: {
      display: 'none',
    },
    color: {
      background: `${color.hex}`,
    },
  };
  const handleClick = () => {
    setColorPickerDisplay(!displayColorPicker);
  };
  const handleClose = () => {
    setColorPickerDisplay(false);
  };
  const handleChange = (color) => {
    if (!color.hex) {
      setColor({hex: initialColor});
      return;
    }
    setColor(color);
    onChange(color, title);
  };
  return (
    <div className="custom-color-picker">
      {title && <div className="custom-color-picker__title">{title}</div>}
      <div className="custom-color-picker__wrapper">
        <div className="custom-color-picker__swatch" style={ styles.swatch } onClick={ handleClick }>
          <div className="custom-color-picker__color" style={ styles.color } />
        </div>
        <EditableInput
          style={ styles }
          label='hex'
          value={ color.hex }
          onChange={ handleChange } />
        { displayColorPicker ? <div className="custom-color-picker__popover" style={ styles.popover }>
            <div className="custom-color-picker__cover" style={ styles.cover } onClick={ handleClose }/>
            <SketchPicker color={ color } onChange={ handleChange } disableAlpha={true} />
          </div> : null }
      </div>
    </div>
  );
}

export default CustomPicker(ColorPicker);

ColorPicker.defaultProps = {
  title: '',
  onChange: () => {},
};

ColorPicker.propTypes = {
  title: PropTypes.string,
  initialColor: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};