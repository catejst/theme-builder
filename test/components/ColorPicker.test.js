import React from 'react'
import { mount } from 'enzyme';
import ColorPicker from '../../lib/ColorPicker';

describe('<ColorPicker />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<ColorPicker />);
  });

  it('ColorPicker component renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should open color picker popover on swatch click', () => {
    wrapper.find('.custom-color-picker__swatch').simulate('click');
    expect(wrapper.find('.custom-color-picker__popover').length).toBe(1);
  });

  it('should close color picker once click on cover', () => {
    wrapper.find('.custom-color-picker__swatch').simulate('click');
    wrapper.find('.custom-color-picker__cover').simulate('click');
    expect(wrapper.find('.custom-color-picker__popover').length).toBe(0);
  });

  describe('props given', () => {
    let wrapper;
    const onChange = jest.fn();
    const title = 'Test';
    const initialColor = 'black';
    const changedColor = 'red';

    beforeEach(() => {
      wrapper = mount(<ColorPicker title={title} initialColor={initialColor} onChange={onChange} />);
    });

    it('should render title', () => {
      expect(wrapper.find('.custom-color-picker__title').text()).toEqual(title);
    });

    it('should render initial color swatch', () => {
      expect(wrapper.find('.custom-color-picker__color').prop('style')).toHaveProperty('background', initialColor);
    });

    it('should render color value in input', () => {
      expect(wrapper.find('input').prop('value')).toEqual(initialColor.toUpperCase());
    });

    it('should call onChange once value is changed', () => {
      wrapper.find('input').simulate('change', { target: { value: changedColor }});
      expect(onChange).toHaveBeenCalled();
    });

    it('should change input value after change', () => {
      wrapper.find('input').simulate('change', { target: { value: changedColor }});
      expect(wrapper.find('input').prop('value')).toEqual(changedColor.toUpperCase());
    });

  });
});