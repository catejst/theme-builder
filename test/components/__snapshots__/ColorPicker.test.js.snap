// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker /> ColorPicker component renders correctly 1`] = `"<div class=\\"custom-color-picker\\"><div class=\\"custom-color-picker__wrapper\\"><div class=\\"custom-color-picker__swatch\\"><div class=\\"custom-color-picker__color\\" style=\\"background: rgb(34, 34, 34);\\"></div></div><div style=\\"position: relative;\\"><input style=\\"outline: none; font-size: 14px;\\" spellcheck=\\"false\\" value=\\"#222222\\"><span style=\\"display: none;\\">hex</span></div></div></div>"`;
